presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java

```diff
@@ -0,0 +1,126 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.Session;
+import io.prestosql.execution.warnings.WarningCollector;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.PlanNodeIdAllocator;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.SymbolAllocator;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.optimizations.PlanOptimizer;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.SimplePlanRewriter;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.Expression;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.Maps.filterKeys;
+import static io.prestosql.sql.DynamicFilters.extractDynamicFilters;
+import static io.prestosql.sql.ExpressionUtils.combineConjuncts;
+import static io.prestosql.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * Dynamic filters are supported only right after TableScan
+ */
+public class RemoveUnsupportedDynamicFilters
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        return SimplePlanRewriter.rewriteWith(new RemoveUnsupportedDynamicFilters.Rewriter(), plan, new HashSet<>());
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Set<String>>
+    {
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Set<String>> context)
+        {
+            JoinNode join = (JoinNode) context.defaultRewrite(node, context.get());
+
+            Optional<Expression> filterOptional = join.getFilter().map(filter -> {
+                Expression modified = removeDynamicFilters(filter, context.get());
+                if (TRUE_LITERAL.equals(modified)) {
+                    return null;
+                }
+                else if (!filter.equals(modified)) {
+                    return modified;
+                }
+                return filter;
+            });
+
+            Map<String, Symbol> dynamicFilters = ImmutableMap.copyOf(filterKeys(node.getDynamicFilters(), key -> !context.get().contains(key)));
```

Comment:
use streams and collectors instead of `filterKeys` and `copyOf`

Reply:
done